coderd/provisionerdserver/provisionerdserver.go

```diff
@@ -1133,23 +1139,42 @@ func (s *server) CompleteJob(ctx context.Context, completed *proto.CompletedJob)
 		for _, externalAuthProvider := range jobType.TemplateImport.ExternalAuthProviders {
 			contains := false
 			for _, configuredProvider := range s.ExternalAuthConfigs {
-				if configuredProvider.ID == externalAuthProvider {
+				if configuredProvider.ID == externalAuthProvider.Id {
 					contains = true
 					break
 				}
 			}
 			if !contains {
 				completedError = sql.NullString{
-					String: fmt.Sprintf("external auth provider %q is not configured", externalAuthProvider),
+					String: fmt.Sprintf("external auth provider %q is not configured", externalAuthProvider.Id),
 					Valid:  true,
 				}
 				break
 			}
 		}
 
+		externalAuthProviders := jobType.TemplateImport.ExternalAuthProviders
+		// Fallback to `ExternalAuthProvidersNames` if it was specified and `ExternalAuthProviders`
+		// was not. Gives us backwards compatibility with custom provisioners that haven't been
+		// updated to use the new field yet.
+		namesLen := len(jobType.TemplateImport.ExternalAuthProvidersNames)
+		if len(externalAuthProviders) == 0 && namesLen > 0 {
+			externalAuthProviders = make([]*sdkproto.ExternalAuthProviderResource, 0, namesLen)
+			for _, providerID := range jobType.TemplateImport.ExternalAuthProvidersNames {
+				externalAuthProviders = append(externalAuthProviders, &sdkproto.ExternalAuthProviderResource{
+					Id: providerID,
+				})
+			}
+		}
+
+		externalAuthProvidersMessage, err := json.Marshal(externalAuthProviders)
```

Comment:
I'm a bit wary of directly serializing an `sdkproto` type to the database, as we don't version that type and kind of treat it as something we can change without consequence.

I think you should define the datatype for `ExternalAuthProviders` column in the `database` package (maybe in `models.go`?) and explicitly convert the `sdkproto` type.

Reply:
models.go is generated by sqlc right? would types.go be ok instead?

Reply:
I think that's a good idea tho, also makes it clearer when the value came from a protobuf message vs when we read it from the database.

Reply:
yeah, types.go is all good; I can never remember which files are generated